src/vfio_device.rs

```diff
@@ -0,0 +1,965 @@
+    GetGroupStatus,
+    GroupViable,
+    VfioApiVersion,
+    VfioExtension,
+    VfioInvalidType,
+    VfioType1V2,
+    GroupSetContainer,
+    UnsetContainer,
+    ContainerSetIOMMU,
+    GroupGetDeviceFD,
+    KvmSetDeviceAttr(SysError),
+    VfioDeviceGetInfo,
+    VfioDeviceGetRegionInfo,
+    InvalidPath,
+    IommuDmaMap,
+    IommuDmaUnmap,
+    VfioDeviceGetIrqInfo,
+    VfioDeviceSetIrq,
+}
+pub type Result<T> = std::result::Result<T, VfioError>;
+
+impl fmt::Display for VfioError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            VfioError::OpenContainer(e) => {
+                write!(f, "failed to open /dev/vfio/vfio container: {}", e)
+            }
+            VfioError::OpenGroup(e) => {
+                write!(f, "failed to open /dev/vfio/$group_num group: {}", e)
+            }
+            VfioError::GetGroupStatus => write!(f, "failed to get Group Status"),
+            VfioError::GroupViable => write!(f, "group is inviable"),
+            VfioError::VfioApiVersion => write!(
+                f,
+                "vfio API version doesn't match with VFIO_API_VERSION defined in vfio-bindings"
+            ),
+            VfioError::VfioExtension => write!(f, "failed to check VFIO extension"),
+            VfioError::VfioInvalidType => write!(f, "invalid VFIO type"),
+            VfioError::VfioType1V2 => {
+                write!(f, "container dones't support VfioType1V2 IOMMU driver type")
+            }
+            VfioError::GroupSetContainer => {
+                write!(f, "failed to add vfio group into vfio container")
+            }
+            VfioError::UnsetContainer => write!(f, "failed to unset vfio container"),
+            VfioError::ContainerSetIOMMU => write!(
+                f,
+                "failed to set container's IOMMU driver type as VfioType1V2"
+            ),
+            VfioError::GroupGetDeviceFD => write!(f, "failed to get vfio device fd"),
+            VfioError::KvmSetDeviceAttr(e) => {
+                write!(f, "failed to set KVM vfio device's attribute: {}", e)
+            }
+            VfioError::VfioDeviceGetInfo => {
+                write!(f, "failed to get vfio device's info or info doesn't match")
+            }
+            VfioError::VfioDeviceGetRegionInfo => {
+                write!(f, "failed to get vfio device's region info")
+            }
+            VfioError::InvalidPath => write!(f, "invalid file path"),
+            VfioError::IommuDmaMap => write!(f, "failed to add guest memory map into iommu table"),
+            VfioError::IommuDmaUnmap => {
+                write!(f, "failed to remove guest memory map from iommu table")
+            }
+            VfioError::VfioDeviceGetIrqInfo => write!(f, "failed to get vfio device irq info"),
+            VfioError::VfioDeviceSetIrq => write!(f, "failed to set vfio deviece irq"),
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Debug, Default)]
+struct vfio_region_info_with_cap {
```

Comment:
Can you add a comment to what this structure represents? You can figure it out from the code, but it takes some time.

Reply:
Done.